Source/Diagnostics/ReducedDiags/ParticleEnergy.cpp

```diff
@@ -0,0 +1,146 @@
+            {
+                ofs << m_sep;
+                ofs << species_names[i]+".mean(J)";
+            }
+            ofs << std::endl;
+            /// close file
+            ofs.close();
+        }
+    }
+
+}
+///< end constructor
+
+/// function that computes kinetic energy
+void ParticleEnergy::ComputeDiags (int step)
+{
+
+    /// Judge if the diags should be done
+    if ( (step+1) % m_freq != 0 ) { return; }
+
+    /// get MultiParticleContainer class object
+    auto & mypc = WarpX::GetInstance().GetPartContainer();
+
+    /// get number of species (int)
+    auto species_number = mypc.nSpecies();
+
+    /// resize data array
+    m_data.resize(2*species_number+2,0.0);
+
+    /// get species names (std::vector<std::string>)
+    auto species_names = mypc.GetSpeciesNames();
+
+    /// speed of light squared
+    auto c2 = PhysConst::c * PhysConst::c;
+
+    /// loop over species
+    for (int i_s = 0; i_s < species_number; ++i_s)
+    {
+        /// get WarpXParticleContainer class object
+        auto & myspc = mypc.GetParticleContainer(i_s);
+
+        /// get mass (Real)
+        auto m = myspc.getMass();
+
+        using PType = typename WarpXParticleContainer::SuperParticleType;
+
+        // Use amex::ReduceSum to compute the sum of energies of all particles
+        // held by the current MPI rank, for this species. This involves a loop over all
+        // boxes held by this MPI rank.
+        auto Etot = ReduceSum( myspc,
+        [=] AMREX_GPU_HOST_DEVICE (const PType& p) -> Real
+        {
+            auto w  = p.rdata(PIdx::w);
+            auto ux = p.rdata(PIdx::ux);
+            auto uy = p.rdata(PIdx::uy);
+            auto uz = p.rdata(PIdx::uz);
+            auto us = (ux*ux + uy*uy + uz*uz);
+            return ( std::sqrt(us*c2 + c2*c2) - c2 ) * m * w;
+        });
+
+        // Same thing for the particles weights.
+        auto Wtot = ReduceSum( myspc,
+        [=] AMREX_GPU_HOST_DEVICE (const PType& p) -> Real
+        {
+            return p.rdata(PIdx::w);
+        });
+
+        /// reduced sum for mpi ranks
+        ParallelDescriptor::ReduceRealSum
+            (Etot, ParallelDescriptor::IOProcessorNumber());
+        ParallelDescriptor::ReduceRealSum
+            (Wtot, ParallelDescriptor::IOProcessorNumber());
+
+        /// save results for this species i_s into m_data
+        m_data[i_s+1] = Etot;
+        if ( Wtot > 0.0 )
+        { m_data[species_number+2+i_s] = Etot / Wtot; }
+        else
+        { m_data[species_number+2+i_s] = 0.0; }
+
+
+    }
+    ///< end loop over species
+
+    /// save total energy
+    /// loop over species
+    m_data[0] = 0.0;                ///< total energy
+    m_data[species_number+1] = 0.0; ///< total mean energy
+    for (int i_s = 0; i_s < species_number; ++i_s)
+    {
+        m_data[0] += m_data[i_s+1];
+        m_data[species_number+1] += m_data[species_number+2+i_s];
+    }
+    ///< end loop over species
```

Comment:
The order in which the quantities are written in `m_data` is hard to determine here. Could you add a line at the end, like
```
// m_data now contains up-to-date values for:
// [total energy (all species), total energy (species 1), ..., total energy (species n),
//  mean  energy (all species), mean  energy (species 1), ..., mean  energy (species n)]
```
or something like that?